### PR TITLE
Add Travis tests with pytest

### PR DESCRIPTION
**Pull request by andrewda**
Closes https://github.com/manu-chroma/username-availability-checker/issues/31

### CODE REVIEW

**Comment by jayvdb** [comment]
Including "invalid_usernames" for each external website is redundant, especially with the data you have.  It is one algorithm, and in that algorithm the API *prevents interaction* with the external website, so it needs only one set of unit tests.  There is no chance of those tests failing because of changes made by the external website.  The invalid username algorithm can be tested directly, hitting the actual functions used.  Then one extra check that invalid usernames respond correctly over HTTP.

"working_usernames" is a confusing name.  It could be "working" correctly as an available or not-available username.   We need lists of "taken usernames" and "available usernames", to check each of the two possible outcomes.  Per the issue description, random usernames are only a fallback for when a known available username is being reported as taken.

Use test data paramaterisation (https://docs.pytest.org/en/latest/parametrize.html ; https://docs.pytest.org/en/latest/example/parametrize.html ; [plugins](https://pypi.python.org/pypi?%3Aaction=search&term=pytest+data&submit=search) ) , to avoid having test methods for each website.  The algorithm for most of them should be identical.  Strings like `'https://soundcloud.com/{}'` need to be moved into data somewhere so they can be accessed generically.

**Comment by jayvdb** [comment]
Also include a link to your travis build each time you want us to do a code review.

@manu-chroma , when you can, could you enable Travis on this repo.

**Comment by manu-chroma** [comment]
done

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+cache: pip
+
+addons:
+  apt:
+    packages:
+      - libssl-dev
+      - python3-dev
+
+before_script:
```
Comment (jayvdb) [inline .travis.yml]: travis automatically installs `requirements.txt` ; trick I use iscat  test-requirements.txt

File: tests/test_username_api.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  def test_taken_twitter_username(self, user):
+    assert_response(self.app, 'twitter', user, 200)
+
+  @pytest.mark.parametrize('user', data['twitter']['available_usernames'])
+  def test_available_github_username(self, user):
+    expected = get_expected_response('github', user, 404)
+    actual = get_response(self.app, 'github', user)
+
+    if expected == actual:
+      assert expected == actual
```
Comment (jayvdb) [inline tests/test_username_api.py]: this should be always run at the bottom of the method, as a human needs to update the `'available_usernames'` data if it fails

add custom fail messages, letting the reader know

1. if a random username also failed to be detected as available.
2. what random username they can use instead of the taken username.

File: .travis.yml (hunk 1 above)
Comment (andrewda) [reply .travis.yml]: Fancy, I like it!

File: tests/test_username_api.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  }
+
+class TestUsernameApi(object):
+  @classmethod
+  def setup_class(self):
+    self.app = username_api.app.test_client()
+    self.data = data
+
+  @pytest.mark.parametrize('user', data['github']['taken_usernames'])
+  def test_taken_github_username(self, user):
```
Comment (jayvdb) [inline tests/test_username_api.py]: the website can also be parameterised, but maybe we leave that for another task.

File: websites.yml [hunk 4]
```diff
@@ -1,36 +1,40 @@
 ---
+urls:
+  behance: https://{w}.net/{u}
+  pinterest: https://in.{w}.com/{u}/
+  tumblr: https://{u}.{w}.com
 username_patterns:
-  pinterest:
```
Comment (jayvdb) [inline websites.yml]: if you want to re-order the data, do it in a separate commit.

**Comment by manu-chroma** [comment]
@andrewda This is much better test than I was expecting. Great work!
